proteus/mprans/SpatialTools.py

```diff
@@ -879,71 +880,124 @@ def __init__(self, domain, dim=(0., 0.), coords=None, from_0=True):
+            self.y0 = 0
+            self.y1 = dim[1]
+        else: # not from_0 and coords is not None
+            self.x0 = coords[0] - 0.5 * dim[0]
+            self.x1 = coords[0] + 0.5 * dim[0]
+            self.y0 = coords[1] - 0.5 * dim[1]
+            self.y1 = coords[1] + 0.5 * dim[1]
+
+    def _constructVertices(self):
+        vertices = [[self.x0, self.y0],
+                    [self.x1, self.y0],
+                    [self.x1, self.y1],
+                    [self.x0, self.y1]]
+        vertexFlags = [self.boundaryTags['y-'],
+                       self.boundaryTags['y-'],
+                       self.boundaryTags['y+'],
+                       self.boundaryTags['y+']]
+        if self.spongeLayers['x-']:
+            vertices += [[self.x0 - self.spongeLayers['x-'], self.y0],
+                         [self.x0 - self.spongeLayers['x-'], self.y1]]
+            vertexFlags += [self.boundaryTags['y-'],
+                            self.boundaryTags['y+']]
+        if self.spongeLayers['x+']:
+            vertices += [[self.x1 + self.spongeLayers['x+'], self.y0],
+                         [self.x1 + self.spongeLayers['x+'], self.y1]]
+            vertexFlags += [self.boundaryTags['y-'],
+                            self.boundaryTags['y+']]
+        return vertices, vertexFlags
+
+    def _constructSegments(self, vertices, vertexFlags):
         segments = [[0, 1], [1, 2], [2, 3], [3, 0]]
-        segmentFlags = [bt['y-'], bt['x+'], bt['y+'], bt['x-']]  # y-, x+, y+, x-
-        regions = [[(x0-leftSponge+x0)/2., regions_y]]
-        regionFlags = [1]
-        self.regionIndice = {'tank': 0}
+        segmentFlags = [self.boundaryTags['y-'],
+                        self.boundaryTags['x+'],
+                        self.boundaryTags['y+'],
+                        self.boundaryTags['x-']]
+        added_vertices = 0
+        if self.spongeLayers['x-']:
+            segments += [[0, 4 + added_vertices],
+                         [4 + added_vertices, 5 + added_vertices],
+                         [5 + added_vertices, 3]]
+            segmentFlags += [self.boundaryTags['y-'],
+                             self.boundaryTags['x-'],
+                             self.boundaryTags['y+']]
+            segmentFlags[3] = self.boundaryTags['sponge']
+            added_vertices += 2
+        if self.spongeLayers['x+']:
+            segments += [[1, 4 + added_vertices],
+                         [4 + added_vertices, 5 + added_vertices],
+                         [5 + added_vertices, 2]]
+            segmentFlags += [self.boundaryTags['y-'],
+                             self.boundaryTags['x+'],
+                             self.boundaryTags['y+']]
+            segmentFlags[1] = self.boundaryTags['sponge']
+            added_vertices += 2
+        return segments, segmentFlags
+
+    def _constructRegions(self, vertices, vertexFlags, segments, segmentFlags):
+        regions = [[0.5 * (self.x0 + self.x1), 0.5 * (self.y0 + self.y1)],]
```

Comment:
For the y coordinates, it might be preferable to put the y (or/and x) coordinates at the top or bottom (or a corner) of the region, where other objects with their own regions are unlikely to clash. In the floating caisson cases, the caisson is usually in the center of the tank for example (I think it is okay though, because it has a "hole", but there could be instances where another object could be in the middle of the tank with its own region).